Fix snapshot manager issues in v3.6

igorsol opened the pull request.

igorsol committed:
use setTimestamp to track possible timestamps in snapshot manager

igorsol committed:
fix snapshot manager

igorsol committed:
fix another RocksSnapshotManager::getCommittedSnapshot call

igorsol committed:
fix more tests

denis-protivenskii commented on src/rocks_snapshot_manager.cpp:
```diff
@@ -92,13 +92,40 @@ namespace mongo {
+    }
+
+    bool RocksSnapshotManager::materializedCommittedSnapshot() const {
+        return _snapshotMap.find(*_committedSnapshot) != _snapshotMap.end();
```
Need `stdx::lock_guard<stdx::mutex> lock(_mutex);` before the return.

denis-protivenskii commented on src/rocks_recovery_unit.cpp:
```diff
@@ -351,6 +357,9 @@ namespace mongo {
         _deltaCounters.clear();
         _writeBatch.Clear();
+        if (_isTimestamped) {
+            _snapshotManager->insertSnapshot(_db, _db->GetSnapshot(), _futureWritesTimestamp);
```
`_isTimestamped = false;`

denis-protivenskii commented on src/rocks_recovery_unit.cpp:
```diff
@@ -373,6 +382,13 @@ namespace mongo {
     }
 
+    std::shared_ptr<RocksSnapshotManager::SnapshotHolder>  RocksRecoveryUnit::_getCommittedSnapshot() const {
+        if (!_snapshotManager->materializedCommittedSnapshot()) {
```
Can you try moving this check into `RocksRecoveryUnit::setReadFromMajorityCommittedSnapshot()`?
I believe this won't break the code (`_getCommittedSnapshot()` is only called after `_readFromMajorityCommittedSnapshot` check right now) and we'll be able to capture majority snapshot as close as possible to the place where it's been requested.

igorsol committed:
add missing mutex lock

igorsol committed:
remove obsolete SnapshotHolder's constructor

igorsol committed:
move materializedCommittedSnapshot call into setReadFromMajorityCommittedSnapshot

igorsol commented:
Now this branch is based on 3.6.1 and the tests are running here: https://jenkins.percona.com/view/PSMDB-3.6/job/percona-server-for-mongodb-3.6-param/14/

igorsol committed:
provide rocks::DB* to rocks snapshot manager

igorsol committed:
do not create snapshots for zero timestamp

igorsol committed:
store snapshots in std::map instead of unoredered_map and vector

in 3.6 secondary nodes replication writers are running in parallel
and can commit recovery units with unordered timestamps
this makes vector maintenance too complex

igorsol committed:
materialize snapshot in setCommittedSnapshot

igorsol committed:
call cleanupUnneededSnapshots() from recovery unit constructor

igorsol committed:
snapshot map optimizations

igorsol committed:
simplify insertSnapshot signature

igorsol committed:
add rocks::DB* parameter to RocksSnapshotManager constructor

denis-protivenskii reviewed (approved):
LGTM
